model/common/src/icon4py/model/common/interpolation/rbf_interpolation.py

```diff
@@ -0,0 +1,520 @@
+# TODO: this is used only in one place, it's probably not as generic as it looks
+def _arc_length_2(
+    v1: data_alloc.NDArray, v2: data_alloc.NDArray, array_ns: ModuleType = np
+) -> data_alloc.NDArray:
+    # For pairs of points p1 and p2 compute:
+    # arccos(dot(p1, p2) / (norm(p1) * norm(p2))) noqa: ERA001
+    # Compute all pairs of dot products
+    arc_lengths = _dot_product(v1, v2, array_ns=array_ns)
+    v1_norm = array_ns.linalg.norm(v1, axis=-1)
+    v2_norm = array_ns.linalg.norm(v2, axis=-1)
+    # Divide the dot products by the broadcasted norms
+    arc_lengths = array_ns.divide(arc_lengths, v1_norm[:, :, array_ns.newaxis])
+    arc_lengths = array_ns.divide(arc_lengths, v2_norm[:, array_ns.newaxis, :])
+    # Ensure all points are within [-1.0, 1.0] (may be outside due to numerical
+    # inaccuracies)
+    arc_lengths = array_ns.clip(arc_lengths, -1.0, 1.0)
+    return array_ns.squeeze(array_ns.arccos(arc_lengths), axis=1)
+
+
+def _gaussian(
+    lengths: data_alloc.NDArray, scale: ta.wpfloat, array_ns: ModuleType = np
+) -> data_alloc.NDArray:
+    val = lengths / scale
+    return array_ns.exp(-1.0 * val * val)
+
+
+def _inverse_multiquadratic(
+    distance: data_alloc.NDArray,
+    scale: ta.wpfloat,
+    array_ns: ModuleType = np,
+) -> data_alloc.NDArray:
+    """
+
+    Args:
+        distance: radial distance
+        scale: scaling parameter
+
+    Returns:
+
+    """
+    val = distance / scale
+    return 1.0 / array_ns.sqrt(1.0 + val * val)
+
+
+def _kernel(
+    kernel: InterpolationKernel,
+    lengths: data_alloc.NDArray,
+    scale: ta.wpfloat,
+    array_ns: ModuleType = np,
+):
+    match kernel:
+        case InterpolationKernel.GAUSSIAN:
+            return _gaussian(lengths, scale, array_ns=array_ns)
+        case InterpolationKernel.INVERSE_MULTIQUADRATIC:
+            return _inverse_multiquadratic(lengths, scale, array_ns=array_ns)
+        case _:
+            raise ValueError(f"Unsupported kernel: {kernel}")
+
+
+def _get_zonal_meridional_f(dim: gtx.Dimension):
+    match dim:
+        case gtx.Dimension("Cell"):
+            return cartesian_coordinates_from_zonal_and_meridional_components_on_cells
+        case gtx.Dimension("Edge"):
+            return cartesian_coordinates_from_zonal_and_meridional_components_on_edges
+        case gtx.Dimension("Vertex"):
+            return cartesian_coordinates_from_zonal_and_meridional_components_on_vertices
+        case _:
+            raise ValueError(f"Unsupported dimension: {dim}")
+
+
+def _compute_rbf_interpolation_matrix(
+    element_center_lat,
+    element_center_lon,
+    element_center_x,
+    element_center_y,
+    element_center_z,
+    edge_center_x,
+    edge_center_y,
+    edge_center_z,
+    edge_normal_x,
+    edge_normal_y,
+    edge_normal_z,
+    u,
+    v,
+    rbf_offset,
+    rbf_kernel: InterpolationKernel,
+    scale_factor: ta.wpfloat,
+    backend: gtx.backend.Backend | None = None,
```

Comment:
you could use the same pattern as in the other functions, that use numpy, pass it as argument. The argument can then be bound by `functools.partial`, you would then have to return `ndarray` s from the function.

Reply:
Updated to only use numpy/cupy and thus pass `array_ns` into the rbf function; previously I had to pass the backend in to get access to some gt4py stuff based on the backend but that's gone now.

I haven't registered the field yet so not doing any `functools.partial` stuff yet, looking at that next.